### PR TITLE
Service provider initial 04 15 22

### PR DESCRIPTION
**Pull request by zapper117**
Added Service Provider pages as per Flow Chart

### CODE REVIEW

File: QuickFixers/Views/ServiceProvider/Index.cshtml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+﻿@{
```
Comment (BSaenz12) [inline QuickFixers/Views/ServiceProvider/Index.cshtml:1]: Please remove dashboard.

File: QuickFixers/Views/Shared/_SPLayout.cshtml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+                <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-collapse">
+                    <span class="icon-bar"></span>
+                    <span class="icon-bar"></span>
+                    <span class="icon-bar"></span>
+                </button>
+                @Html.ActionLink("QuickFix", "Index", "ServiceProvider", new { area = "" }, new { @class = "navbar-brand" })
+            </div>
+            <div class="navbar-collapse collapse">
+                <ul class="nav navbar-nav">
+                    <li>@Html.ActionLink("Dashboard", "Index", "ServiceProvider")</li>
```
Comment (BSaenz12) [inline QuickFixers/Views/Shared/_SPLayout.cshtml]: Please remove Dashboard

File: QuickFixers/Controllers/ServiceProviderController.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+            ViewBag.Message = "Shows past services and clickable action to view an associated survey with each service";
+
+            return View();
+        }
+
+        public ActionResult LogOut()
+        {
+            ViewBag.Message = "Logs out user and redirects to the QuickFix main page";
+
+            return View();
```
Comment (Jody-Whitis) [inline QuickFixers/Controllers/ServiceProviderController.cs]: Please review if we'll need the logout logic from HomeController/Logout here too.

File: QuickFixers/Views/Shared/_SPLayout.cshtml [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+                    <li>@Html.ActionLink("Work Schedule", "WorkSchedule", "ServiceProvider")</li>
+                    <li>@Html.ActionLink("Invoices", "Invoices", "ServiceProvider")</li>
+                    <li>@Html.ActionLink("Services Offered", "ServicesOffered", "ServiceProvider")</li>
+                    <li>@Html.ActionLink("Scheduled Services", "ScheduledServices", "ServiceProvider")</li>
+                    <li>@Html.ActionLink("Past Services", "PastServices", "ServiceProvider")</li>
+                    <li>@Html.ActionLink("Log Out", "LogOut", "ServiceProvider")</li>
```
Comment (Jody-Whitis) [inline QuickFixers/Views/Shared/_SPLayout.cshtml]: Shared/_Layout should already have logout logic in Controllers/HomeController.cs/Logout. Please remove all logout functions and confirm if we can reuse the existing Logout function.

image

image

File: QuickFixers/Controllers/LoginController.cs [hunk 5]
```diff
@@ -79,8 +88,18 @@ public ActionResult Register(LoginViewModel newLoginViewModel)
         {
             if (ModelState.IsValid)
             {
-                IUser newUser = newLoginViewModel.UserTypeID == 1 ? new Clients() : null; //replace null with service provider
-                
+                //IUser newUser = newLoginViewModel.UserTypeID == 1 ? new Clients() : null; //replace null with service provider
```
Comment (BSaenz12) [inline QuickFixers/Controllers/LoginController.cs]: Remove commented code.

File: QuickFixers/Controllers/ServiceProviderController.cs [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+
+namespace QuickFixers.Controllers
+{
+    public class ServiceProviderController : Controller
+    {
+        // GET: SP
+        public ActionResult Index()
+        {
+            //Manually Set SeviceProvider Info to allow for building of pages
```
Comment (BSaenz12) [inline QuickFixers/Controllers/ServiceProviderController.cs]: Remove commented code.

File: QuickFixers/Controllers/ServiceProviderController.cs [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+
+            return View();
+        }
+
+        public ActionResult WorkSchedule()
+        {
+            ServiceProviderViewModel wsresults = new ServiceProviderViewModel();
+            wsresults.DbResults = Data.DataBase.DatabaseSelections.SelectSPWorkSchedule((int)Session["ServiceProviderID"]);
+            wsresults.IsDBConnected = wsresults.DbResults != null && wsresults.DbResults.Rows.Count > 0 ? true : false;
+            //return View(wsresults);
```
Comment (BSaenz12) [inline QuickFixers/Controllers/ServiceProviderController.cs]: Remove commented code.

File: QuickFixers/Controllers/ServiceProviderController.cs [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+
+            return View(wsresults);
+
+            //return View();
+        }
+
+        public ActionResult Invoices()
+        {
+
+            ServiceProviderViewModel invoiceresults = new ServiceProviderViewModel();
```
Comment (BSaenz12) [inline QuickFixers/Controllers/ServiceProviderController.cs:58]: Remove commented code

File: QuickFixers/Views/ServiceProvider/Invoices.cshtml [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+            <tr>
+                @{
+
+                    foreach (var header in THeaders)
+                    {
+
+                        <th>
+                            @header
+                        </th>
+
```
Comment (BSaenz12) [inline QuickFixers/Views/ServiceProvider/Invoices.cshtml]: Removed unnecessary lines.

File: QuickFixers/Views/ServiceProvider/ServicesOffered.cshtml [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+            <tr>
+                @{
+
+                    foreach (var header in THeaders)
+                    {
+
+                        <th>
+                            @header
+                        </th>
+
```
Comment (BSaenz12) [inline QuickFixers/Views/ServiceProvider/ServicesOffered.cshtml]: Remove unnecessary lines.

File: QuickFixers/Views/ServiceProvider/WorkSchedule.cshtml [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+            <tr>
+                @{
+
+                    foreach (var header in THeaders)
+                    {
+
+                        <th>
+                            @header
+                        </th>
+
```
Comment (BSaenz12) [inline QuickFixers/Views/ServiceProvider/WorkSchedule.cshtml]: Remove empty lines.

File: QuickFixers/Views/Shared/_SPLayout.cshtml [hunk 12]
```diff
@@ -0,0 +1,49 @@
+﻿<!DOCTYPE html>
+<html>
```
Comment (BSaenz12) [inline QuickFixers/Views/Shared/_SPLayout.cshtml:2]: Work with @BSaenz12 to resolve nav bar logic and conflicts.

File: QuickFixers/Views/Test/Index.cshtml [hunk 13]
```diff
@@ -1,6 +1,7 @@
 ﻿@model QuickFixers.Models.TestViewModel
 
 <h2>Is DataBase Connected: @Model.IsDBConnected</h2>
+
```
Comment (BSaenz12) [inline QuickFixers/Views/Test/Index.cshtml]: Remove empty line.

File: QuickFixers.Data/DataBase/DatabaseSelections.cs [hunk 14]
```diff
@@ -70,5 +70,190 @@ public static DataTable SelectUser(User loggingUser)
...
+            {
+                MySqlConnectionStringBuilder connectionStringBuilder = new MySqlConnectionStringBuilder();
+                connectionStringBuilder = DatabaseExtensions.ToConnectionStringBuilder(connectionStringBuilder);
+
+                using (var connectionDB = new MySqlConnection(connectionStringBuilder.ToString()))
+                {
+                    using (var sqlQuery = new MySqlCommand("quickFixers.SelSPUser", connectionDB))
+                    {
+                        sqlQuery.CommandType = CommandType.StoredProcedure;
+                        sqlQuery.Parameters.AddWithValue($"@UsrID", sessionUserID);
```
Comment (BSaenz12) [inline QuickFixers.Data/DataBase/DatabaseSelections.cs]: Please update parameter name to `@UserID`

File: QuickFixers.Data/DataBase/DatabaseSelections.cs [hunk 15]
```diff
@@ -70,5 +70,190 @@ public static DataTable SelectUser(User loggingUser)
...
+                        //sqlQuery.Parameters.AddWithValue($"@UserPassword", loggingUser.UserPassword);
```
Comment (BSaenz12) [inline QuickFixers.Data/DataBase/DatabaseSelections.cs]: Remove commented code.

File: QuickFixers.Data/DataBase/DatabaseSelections.cs [hunk 16]
```diff
@@ -70,5 +70,190 @@ public static DataTable SelectUser(User loggingUser)
...
+            {
+                returnResults = new DataTable();
+                Console.WriteLine(ex.ToString());
+            }
+            return returnResults;
+        }
+
+
+        //Used to retrieve Service Provider Work Schedule Data Table for Specified Service Provider
+        public static DataTable SelectSPWorkSchedule(int SPUsrID)
```
Comment (BSaenz12) [inline QuickFixers.Data/DataBase/DatabaseSelections.cs]: Please rename parameter to `SPUserID` in app code and db.

File: QuickFixers.Data/DataBase/DatabaseSelections.cs [hunk 17]
```diff
@@ -70,5 +70,190 @@ public static DataTable SelectUser(User loggingUser)
...
+                MySqlConnectionStringBuilder connectionStringBuilder = new MySqlConnectionStringBuilder();
+                connectionStringBuilder = DatabaseExtensions.ToConnectionStringBuilder(connectionStringBuilder);
+
+                using (var connectionDB = new MySqlConnection(connectionStringBuilder.ToString()))
+                {
+                    using (var sqlQuery = new MySqlCommand("quickFixers.SelSPWorkSchedule", connectionDB))
+                    {
+                        sqlQuery.CommandType = CommandType.StoredProcedure;
+                        sqlQuery.Parameters.AddWithValue($"@SPID", SPUsrID);
+                        //sqlQuery.Parameters.AddWithValue($"@UserPassword", loggingUser.UserPassword);
```
Comment (BSaenz12) [inline QuickFixers.Data/DataBase/DatabaseSelections.cs]: Remove commented code.

File: QuickFixers.Data/DataBase/DatabaseSelections.cs [hunk 18]
```diff
@@ -70,5 +70,190 @@ public static DataTable SelectUser(User loggingUser)
...
+                MySqlConnectionStringBuilder connectionStringBuilder = new MySqlConnectionStringBuilder();
+                connectionStringBuilder = DatabaseExtensions.ToConnectionStringBuilder(connectionStringBuilder);
+
+                using (var connectionDB = new MySqlConnection(connectionStringBuilder.ToString()))
+                {
+                    using (var sqlQuery = new MySqlCommand("quickFixers.SelSPServicesOffered", connectionDB))
+                    {
+                        sqlQuery.CommandType = CommandType.StoredProcedure;
+                        sqlQuery.Parameters.AddWithValue($"@SPID", SPUsrID);
+                        //sqlQuery.Parameters.AddWithValue($"@UserPassword", loggingUser.UserPassword);
```
Comment (BSaenz12) [inline QuickFixers.Data/DataBase/DatabaseSelections.cs]: Remove commented code

File: QuickFixers.Data/DataBase/DatabaseSelections.cs [hunk 19]
```diff
@@ -70,5 +70,190 @@ public static DataTable SelectUser(User loggingUser)
...
+            catch (Exception ex)
+            {
+                returnResults = new DataTable();
+                Console.WriteLine(ex.ToString());
+            }
+            return returnResults;
+        }
+
+        //Used to retrieve Service Provider Work Schedule Data Table for Specified Service Provider
+        public static DataTable SelectSPServiceOffered(int SPUsrID)
```
Comment (BSaenz12) [inline QuickFixers.Data/DataBase/DatabaseSelections.cs]: Rename parameter to `SPUserID`

File: QuickFixers.Data/DataBase/DatabaseSelections.cs (hunk 14 above)
Comment (zapper117) [reply QuickFixers.Data/DataBase/DatabaseSelections.cs]: Updated to UserID in App code and db Stored Procedure

File: QuickFixers/Views/Shared/_SPLayout.cshtml (hunk 12 above)
Comment (zapper117) [reply QuickFixers/Views/Shared/_SPLayout.cshtml:2]: In-Progress...

File: QuickFixers.Data/Models/Clients.cs [hunk 20]
```diff
@@ -8,6 +8,8 @@ namespace QuickFixers.Data.Models
 {
    public class Clients: User, IUser
     {
+
+        private int defaultClientID = -1;
```
Comment (BSaenz12) [inline QuickFixers.Data/Models/Clients.cs]: Why is this set `-1` ?

File: QuickFixers.Data/Models/ServiceProvider.cs [hunk 21]
```diff
@@ -0,0 +1,22 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+namespace QuickFixers.Data.Models
+{
+    public class ServiceProvider: User, IUser
+    {
+        public int ServiceProviderID { get; set; }
+        //Set Default ClientID for IUser
+        private int defaultClientID = -1;
```
Comment (BSaenz12) [inline QuickFixers.Data/Models/ServiceProvider.cs:13]: Why is this set to `-1` ?

File: QuickFixers/Controllers/ServiceProviderController.cs [hunk 22]
```diff
@@ -0,0 +1,111 @@
...
+        {
+
+            ServiceProviderViewModel scheduledresults = new ServiceProviderViewModel();
+            scheduledresults.DbResults = Data.DataBase.DatabaseSelections.SelectSPScheduledServices((int)Session["ServiceProviderID"]);
+            scheduledresults.IsDBConnected = scheduledresults.DbResults != null && scheduledresults.DbResults.Rows.Count > 0 ? true : false;
+
+            ViewBag.Message = "Shows current scheduled services ";
+
+            return View(scheduledresults);
+
```
Comment (BSaenz12) [inline QuickFixers/Controllers/ServiceProviderController.cs:91]: Remove empty lines.

File: QuickFixers.Data/DataBase/DatabaseSelections.cs [hunk 23]
```diff
@@ -70,5 +70,184 @@ public static DataTable SelectUser(User loggingUser)
             }
             return returnResults;
         }
+
+        //Used to retrieve Service Provider info and save to Session for later use
+        public static DataTable SelectSP(int sessionUserID)
```
Comment (BSaenz12) [inline QuickFixers.Data/DataBase/DatabaseSelections.cs:108]: Please replace the select functions with the new generic select function.

**Review by BSaenz12 — APPROVED** [review]
Good work

**Pull request merged**